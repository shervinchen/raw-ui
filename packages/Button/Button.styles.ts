import css from 'styled-jsx/css';
import { RawUITheme } from '../Theme/preset/preset.type';
import { useTheme } from '../Theme/theme-context';
import {
  ButtonSizes,
  ButtonTypes,
  ButtonVariants,
  ButtonProps,
  ButtonSizeStyles,
  ButtonBasicStyles,
  ButtonStyles,
  ButtonHoverStyles,
  ButtonActiveStyles,
} from './Button.types';

export const useButtonStyles = (props: ButtonProps): ButtonStyles => {
  const theme: RawUITheme = useTheme();
  const { size, type, variant, loading, disabled } = props;

  const sizes: {
    [key in ButtonSizes]: ButtonSizeStyles;
  } = {
    xs: {
      fontSize: '12px',
      height: '24px',
      horizontalPadding: '12px',
    },
    sm: {
      fontSize: '14px',
      height: '28px',
      horizontalPadding: '16px',
    },
    md: {
      fontSize: '14px',
      height: '32px',
      horizontalPadding: '16px',
    },
    lg: {
      fontSize: '14px',
      height: '36px',
      horizontalPadding: '16px',
    },
    xl: {
      fontSize: '16px',
      height: '40px',
      horizontalPadding: '20px',
    },
  };

  const styles: {
    [key in ButtonTypes]: {
      [key in ButtonVariants]: ButtonBasicStyles;
    };
  } = {
    default: {
      default: {
        backgroundColor: theme.palette.background,
        color: theme.palette.neutral['600'],
        borderColor: theme.palette.neutral['200'],
      },
      outline: {
        backgroundColor: theme.palette.background,
        color: theme.palette.neutral['600'],
        borderColor: theme.palette.neutral['200'],
      },
      ghost: {
        backgroundColor: 'transparent',
        color: theme.palette.neutral['600'],
        borderColor: 'transparent',
      },
      shadow: {
        backgroundColor: theme.palette.background,
        color: theme.palette.neutral['600'],
        borderColor: theme.palette.background,
        boxShadow: theme.tokens.shadow.sm,
      },
    },
    primary: {
      default: {
        backgroundColor: theme.palette.foreground,
        color: theme.palette.background,
        borderColor: theme.palette.foreground,
      },
      outline: {
        backgroundColor: 'transparent',
        color: theme.palette.foreground,
        borderColor: theme.palette.foreground,
      },
      ghost: {
        backgroundColor: 'transparent',
        color: theme.palette.foreground,
        borderColor: 'transparent',
      },
      shadow: {
        backgroundColor: theme.palette.foreground,
        color: theme.palette.background,
        borderColor: theme.palette.foreground,
        boxShadow: theme.tokens.shadow.sm,
      },
    },
    success: {
      default: {
        backgroundColor: theme.palette.blue['500'],
        color: theme.palette.background,
        borderColor: theme.palette.blue['500'],
      },
      outline: {
        backgroundColor: 'transparent',
        color: theme.palette.blue['500'],
        borderColor: theme.palette.blue['500'],
      },
      ghost: {
        backgroundColor: 'transparent',
        color: theme.palette.blue['500'],
        borderColor: 'transparent',
      },
      shadow: {
        backgroundColor: theme.palette.blue['500'],
        color: theme.palette.background,
        borderColor: theme.palette.blue['500'],
        boxShadow: theme.tokens.shadow.sm,
      },
    },
    warning: {
      default: {
        backgroundColor: theme.palette.yellow['500'],
        color: theme.palette.background,
        borderColor: theme.palette.yellow['500'],
      },
      outline: {
        backgroundColor: 'transparent',
        color: theme.palette.yellow['500'],
        borderColor: theme.palette.yellow['500'],
      },
      ghost: {
        backgroundColor: 'transparent',
        color: theme.palette.yellow['500'],
        borderColor: 'transparent',
      },
      shadow: {
        backgroundColor: theme.palette.yellow['500'],
        color: theme.palette.background,
        borderColor: theme.palette.yellow['500'],
        boxShadow: theme.tokens.shadow.sm,
      },
    },
    error: {
      default: {
        backgroundColor: theme.palette.red['500'],
        color: theme.palette.background,
        borderColor: theme.palette.red['500'],
      },
      outline: {
        backgroundColor: 'transparent',
        color: theme.palette.red['500'],
        borderColor: theme.palette.red['500'],
      },
      ghost: {
        backgroundColor: 'transparent',
        color: theme.palette.red['500'],
        borderColor: 'transparent',
      },
      shadow: {
        backgroundColor: theme.palette.red['500'],
        color: theme.palette.background,
        borderColor: theme.palette.red['500'],
        boxShadow: theme.tokens.shadow.sm,
      },
    },
  };

  const loadingStyles: {
    [key in ButtonTypes]: {
      [key in ButtonVariants]: ButtonBasicStyles;
    };
  } = {
    default: {
      default: {
        backgroundColor: theme.palette.background,
      },
      outline: {
        backgroundColor: theme.palette.background,
      },
      ghost: {
        backgroundColor: theme.palette.background,
      },
      shadow: {
        backgroundColor: theme.palette.background,
      },
    },
    primary: {
      default: {
        backgroundColor: theme.palette.foreground,
      },
      outline: {
        backgroundColor: theme.palette.background,
      },
      ghost: {
        backgroundColor: theme.palette.background,
      },
      shadow: {
        backgroundColor: theme.palette.foreground,
      },
    },
    success: {
      default: {
        backgroundColor: theme.palette.blue['500'],
      },
      outline: {
        backgroundColor: theme.palette.background,
      },
      ghost: {
        backgroundColor: theme.palette.background,
      },
      shadow: {
        backgroundColor: theme.palette.blue['500'],
      },
    },
    warning: {
      default: {
        backgroundColor: theme.palette.yellow['500'],
      },
      outline: {
        backgroundColor: theme.palette.background,
      },
      ghost: {
        backgroundColor: theme.palette.background,
      },
      shadow: {
        backgroundColor: theme.palette.yellow['500'],
      },
    },
    error: {
      default: {
        backgroundColor: theme.palette.red['500'],
      },
      outline: {
        backgroundColor: theme.palette.background,
      },
      ghost: {
        backgroundColor: theme.palette.background,
      },
      shadow: {
        backgroundColor: theme.palette.red['500'],
      },
    },
  };

  const disabledStyles: {
    [key in ButtonVariants]: ButtonBasicStyles;
  } = {
    default: {
      backgroundColor: theme.palette.neutral['200'],
      color: theme.palette.neutral['400'],
      borderColor: theme.palette.neutral['200'],
    },
    outline: {
      backgroundColor: theme.palette.neutral['50'],
      color: theme.palette.neutral['400'],
      borderColor: theme.palette.neutral['200'],
    },
    ghost: {
      color: theme.palette.neutral['400'],
    },
    shadow: {
      backgroundColor: theme.palette.neutral['200'],
      color: theme.palette.neutral['400'],
      borderColor: theme.palette.neutral['200'],
    },
  };

  const defaultStyles = {
    ...(styles?.[type || 'default']?.[variant || 'default'] ??
      styles['default']['default']),
    ...(sizes?.[size || 'md'] ?? sizes['md']),
  };

  return {
    ...defaultStyles,
    ...(loading
      ? {
          ...(loadingStyles?.[type || 'default']?.[variant || 'default'] ??
            loadingStyles['default']['default']),
          cursor: 'default',
        }
      : {}),
    ...(disabled
      ? {
          ...(type === 'default'
            ? disabledStyles['outline']
            : disabledStyles?.[variant || 'default'] ??
              disabledStyles['default']),
          cursor: 'not-allowed',
        }
      : {}),
  };
};

export const useButtonHoverStyles = (props: ButtonProps): ButtonHoverStyles => {
  const theme: RawUITheme = useTheme();
  const { type, variant, loading, disabled } = props;
  const styles: {
    [key in ButtonTypes]: {
      [key in ButtonVariants]: ButtonHoverStyles;
    };
  } = {
    default: {
      default: {
        hoverColor: theme.palette.foreground,
        hoverBorderColor: theme.palette.foreground,
      },
      outline: {
        hoverColor: theme.palette.foreground,
        hoverBorderColor: theme.palette.foreground,
      },
      ghost: {
        hoverBackgroundColor: theme.palette.neutral['200'],
        hoverColor: theme.palette.neutral['600'],
      },
      shadow: {
        hoverBoxShadow: theme.tokens.shadow.md,
        hoverTransform: 'translateY(-2px)',
      },
    },
    primary: {
      default: {
        hoverColor: theme.palette.foreground,
        hoverBackgroundColor: theme.palette.background,
      },
      outline: {
        hoverColor: theme.palette.background,
        hoverBackgroundColor: theme.palette.foreground,
      },
      ghost: {
        hoverBackgroundColor: theme.palette.neutral['200'],
      },
      shadow: {
        hoverBoxShadow: theme.tokens.shadow.md,
        hoverTransform: 'translateY(-2px)',
      },
    },
    success: {
      default: {
        hoverColor: theme.palette.blue['500'],
        hoverBackgroundColor: theme.palette.background,
      },
      outline: {
        hoverColor: theme.palette.background,
        hoverBackgroundColor: theme.palette.blue['500'],
      },
      ghost: {
        hoverBackgroundColor: theme.palette.blue['100'],
      },
      shadow: {
        hoverBoxShadow: theme.tokens.shadow.md,
        hoverTransform: 'translateY(-2px)',
      },
    },
    warning: {
      default: {
        hoverColor: theme.palette.yellow['500'],
        hoverBackgroundColor: theme.palette.background,
      },
      outline: {
        hoverColor: theme.palette.background,
        hoverBackgroundColor: theme.palette.yellow['500'],
      },
      ghost: {
        hoverBackgroundColor: theme.palette.yellow['100'],
      },
      shadow: {
        hoverBoxShadow: theme.tokens.shadow.md,
        hoverTransform: 'translateY(-2px)',
      },
    },
    error: {
      default: {
        hoverColor: theme.palette.red['500'],
        hoverBackgroundColor: theme.palette.background,
      },
      outline: {
        hoverColor: theme.palette.background,
        hoverBackgroundColor: theme.palette.red['500'],
      },
      ghost: {
        hoverBackgroundColor: theme.palette.red['100'],
      },
      shadow: {
        hoverBoxShadow: theme.tokens.shadow.md,
        hoverTransform: 'translateY(-2px)',
      },
    },
  };

  const defaultStyles =
    styles?.[type || 'default']?.[variant || 'default'] ??
    styles['default']['default'];

  return loading || disabled ? {} : defaultStyles;
};

export const useButtonActiveStyles = (
  props: ButtonProps
): ButtonActiveStyles => {
  const theme: RawUITheme = useTheme();
  const { type, variant, loading, disabled } = props;
  const styles: {
    [key in ButtonTypes]: {
      [key in ButtonVariants]: ButtonActiveStyles;
    };
  } = {
    default: {
      default: {
        activeBackgroundColor: theme.palette.neutral['200'],
      },
      outline: {
        activeBackgroundColor: theme.palette.neutral['200'],
      },
      ghost: {
        activeBackgroundColor: theme.palette.neutral['300'],
      },
      shadow: {
        activeBoxShadow: theme.tokens.shadow.sm,
        activeTransform: 'none',
      },
    },
    primary: {
      default: {
        activeBackgroundColor: theme.palette.neutral['200'],
      },
      outline: {
        activeBackgroundColor: theme.palette.neutral['800'],
      },
      ghost: {
        activeBackgroundColor: theme.palette.neutral['300'],
      },
      shadow: {
        activeBoxShadow: theme.tokens.shadow.sm,
        activeTransform: 'none',
      },
    },
    success: {
      default: {
        activeBackgroundColor: theme.palette.blue['200'],
      },
      outline: {
        activeBackgroundColor: theme.palette.blue['400'],
      },
      ghost: {
        activeBackgroundColor: theme.palette.blue['200'],
      },
      shadow: {
        activeBoxShadow: theme.tokens.shadow.sm,
        activeTransform: 'none',
      },
    },
    warning: {
      default: {
        activeBackgroundColor: theme.palette.yellow['200'],
      },
      outline: {
        activeBackgroundColor: theme.palette.yellow['400'],
      },
      ghost: {
        activeBackgroundColor: theme.palette.yellow['200'],
      },
      shadow: {
        activeBoxShadow: theme.tokens.shadow.sm,
        activeTransform: 'none',
      },
    },
    error: {
      default: {
        activeBackgroundColor: theme.palette.red['200'],
      },
      outline: {
        activeBackgroundColor: theme.palette.red['400'],
      },
      ghost: {
        activeBackgroundColor: theme.palette.red['200'],
      },
      shadow: {
        activeBoxShadow: theme.tokens.shadow.sm,
        activeTransform: 'none',
      },
    },
  };

  const defaultStyles =
    styles?.[type || 'default']?.[variant || 'default'] ??
    styles['default']['default'];

  return loading || disabled ? {} : defaultStyles;
};

export const useButtonCSS = (props: ButtonProps) => {
  const {
    fontSize,
    height,
    horizontalPadding,
    backgroundColor,
    borderColor,
    color,
    boxShadow = 'none',
    cursor = 'pointer',
  } = useButtonStyles(props);

  const {
    hoverBackgroundColor = backgroundColor,
    hoverBorderColor = borderColor,
    hoverColor = color,
    hoverBoxShadow = 'none',
    hoverTransform = 'none',
  } = useButtonHoverStyles(props);

  const {
    activeBackgroundColor = backgroundColor,
    activeBoxShadow = 'none',
    activeTransform = 'none',
  } = useButtonActiveStyles(props);

  return css.resolve`
    .raw-button:not(.raw-loading-button):not(.raw-disabled-button):hover,
    .raw-button:not(.raw-loading-button):not(.raw-disabled-button):focus {
      background-color: ${hoverBackgroundColor};
      border-color: ${hoverBorderColor};
      color: ${hoverColor};
      box-shadow: ${hoverBoxShadow};
      transform: ${hoverTransform};
    }

    .raw-button:not(.raw-loading-button):not(.raw-disabled-button):active {
      background-color: ${activeBackgroundColor};
      box-shadow: ${activeBoxShadow};
      transform: ${activeTransform};
    }

    .raw-button {
      box-sizing: border-box;
      position: relative;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex-shrink: 0;
      vertical-align: middle;
      border: 1px solid;
      border-radius: 6px;
      font-weight: 400;
      font-size: ${fontSize};
      line-height: 1;
      height: ${height};
      padding: 0 ${horizontalPadding};
      background-color: ${backgroundColor};
      border-color: ${borderColor};
      color: ${color};
      box-shadow: ${boxShadow};
      cursor: ${cursor};
      transition-property: border-color, background, color, transform,
        box-shadow;
      transition-duration: 0.2s;
      transition-timing-function: ease;
      text-transform: capitalize;
      white-space: nowrap;
      appearance: none;
      outline: none;
      user-select: none;
    }

    .raw-button.raw-childless-button {
      width: ${height};
      height: ${height};
      padding: 0;
    }

    .raw-button :global(.raw-button-content) {
      position: relative;
      z-index: 1;
    }
  `;
};
