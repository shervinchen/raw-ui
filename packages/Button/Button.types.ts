import { MouseEventHandler, ButtonHTMLAttributes, ReactNode } from 'react';

export type ButtonSizes = 'sm' | 'md' | 'lg';

export type ButtonTypes =
  | 'default'
  | 'primary'
  | 'success'
  | 'warning'
  | 'error';

export type ButtonVariants = 'default' | 'outline' | 'ghost' | 'shadow';

export interface BaseButtonProps {
  className?: string;
  size?: ButtonSizes;
  type?: ButtonTypes;
  variant?: ButtonVariants;
  htmlType?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
  loading?: boolean;
  disabled?: boolean;
  icon?: ReactNode;
  iconRight?: ReactNode;
  onClick?: MouseEventHandler<HTMLButtonElement>;
}

type NativeButtonProps = Omit<
  ButtonHTMLAttributes<HTMLButtonElement>,
  keyof BaseButtonProps
>;

export type ButtonProps = BaseButtonProps & NativeButtonProps;

export interface ButtonLoadingProps {
  color?: string;
  backgroundColor?: string;
}

export interface ButtonIconProps {
  isRight?: boolean;
  isSingle?: boolean;
  height?: string;
  horizontalPadding?: string;
}

export interface ButtonSizeStyles {
  fontSize?: string;
  lineHeight?: string;
  height?: string;
  horizontalPadding?: string;
}

export interface ButtonBasicStyles {
  backgroundColor?: string;
  borderColor?: string;
  color?: string;
  boxShadow?: string;
  cursor?: 'pointer' | 'default' | 'not-allowed';
}

export type ButtonStyles = ButtonSizeStyles & ButtonBasicStyles;

export interface ButtonHoverStyles {
  hoverBackgroundColor?: string;
  hoverBorderColor?: string;
  hoverColor?: string;
  hoverBoxShadow?: string;
  hoverTransform?: string;
}

export interface ButtonActiveStyles {
  activeBackgroundColor?: string;
  activeBoxShadow?: string;
  activeTransform?: string;
}
