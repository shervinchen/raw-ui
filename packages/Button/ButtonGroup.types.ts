import { HTMLAttributes } from 'react';
import { ButtonSizes, ButtonTypes } from './Button.types';

export type ButtonGroupVariant = 'default' | 'outline' | 'ghost';

interface BaseButtonGroupProps {
  className?: string;
  size?: ButtonSizes;
  type?: ButtonTypes;
  variant?: ButtonGroupVariant;
  disabled?: boolean;
  vertical?: boolean;
}

type NativeButtonGroupProps = Omit<
  HTMLAttributes<HTMLDivElement>,
  keyof BaseButtonGroupProps
>;

export type ButtonGroupProps = BaseButtonGroupProps & NativeButtonGroupProps;

export interface ButtonGroupConfig {
  size?: ButtonSizes;
  type?: ButtonTypes;
  variant?: ButtonGroupVariant;
  disabled?: boolean;
  isButtonGroup: boolean;
}
