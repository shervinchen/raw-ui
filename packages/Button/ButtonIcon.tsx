import React, { FC, PropsWithChildren } from 'react';
import css from 'styled-jsx/css';
import classNames from 'classnames';

import { ButtonIconProps } from './Button.types';

const ButtonIcon: FC<PropsWithChildren<ButtonIconProps>> = ({
  isRight = false,
  isSingle = false,
  height,
  horizontalPadding,
  children,
}) => {
  const { className, styles } = css.resolve`
    .raw-button-icon {
      display: flex;
      justify-content: center;
      align-items: center;
      z-index: 1;
      margin-right: calc(${horizontalPadding} / 2);
    }

    .raw-button-icon :global(svg) {
      background: transparent;
      height: calc(${height} / 2.35);
      width: calc(${height} / 2.35);
    }

    .button-icon-right {
      margin-left: calc(${horizontalPadding} / 2);
      margin-right: 0;
    }

    .button-icon-single {
      margin-right: 0;
    }
  `;

  const classes = classNames(
    'raw-button-icon',
    {
      'button-icon-right': isRight,
      'button-icon-single': isSingle,
    },
    className
  );

  return (
    <span className={classes}>
      {children}
      {styles}
    </span>
  );
};

export default ButtonIcon;
