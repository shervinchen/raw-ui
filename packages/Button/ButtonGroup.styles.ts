import css from 'styled-jsx/css';
import { useButtonHoverStyles, useButtonStyles } from './Button.styles';
import { ButtonGroupProps } from './ButtonGroup.types';

export const useButtonGroupCSS = (props: ButtonGroupProps) => {
  const { type, size, variant, disabled } = props;

  const { borderColor } = useButtonStyles({
    type,
    size,
    variant,
    disabled,
  });

  const { hoverBorderColor } = useButtonHoverStyles({
    type,
    size,
    variant,
    disabled,
  });

  return css.resolve`
    .raw-button-group {
      display: inline-flex;
      background-color: transparent;
      border-radius: 6px;
      overflow: hidden;
    }
    .raw-button-group :global(.raw-button) {
      border: none;
      backface-visibility: hidden;
      transform: translate3d(0, 0, 0);
    }
    .raw-button-group :global(.raw-button::before),
    .raw-button-group :global(.raw-button::after) {
      position: absolute;
      left: 0;
      top: 0;
      right: 0;
      bottom: 0;
      content: '';
      border-radius: inherit;
      transition: border-color 0.2s ease;
      pointer-events: none;
    }
    .raw-button-group :global(.raw-button::before) {
      border: 1px solid ${borderColor};
      z-index: 3;
    }
    .raw-button-group :global(.raw-button::after) {
      border: 1px solid transparent;
      z-index: 3;
    }
    .raw-button-group
      :global(.raw-button:not(.raw-disabled-button):hover::after),
    .raw-button-group
      :global(.raw-button:not(.raw-disabled-button):focus::after) {
      border: 1px solid ${hoverBorderColor};
    }
    .horizontal :global(.raw-button:not(:first-child)) {
      border-top-left-radius: 0;
      border-bottom-left-radius: 0;
    }
    .horizontal :global(.raw-button:not(:last-child)) {
      border-top-right-radius: 0;
      border-bottom-right-radius: 0;
    }
    .horizontal :global(.raw-button:not(:first-child)::before) {
      border-left-width: 0;
    }
    .horizontal :global(.raw-button:not(:first-child)::after) {
      left: -1px;
    }
    .vertical {
      flex-direction: column;
    }
    .vertical :global(.raw-button:not(:first-child)) {
      border-top-left-radius: 0;
      border-top-right-radius: 0;
    }
    .vertical :global(.raw-button:not(:last-child)) {
      border-bottom-left-radius: 0;
      border-bottom-right-radius: 0;
    }
    .vertical :global(.raw-button:not(:first-child)::before) {
      border-top-width: 0;
    }
    .vertical :global(.raw-button:not(:first-child)::after) {
      top: -1px;
    }
  `;
};
