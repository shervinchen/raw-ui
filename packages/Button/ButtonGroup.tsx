import React, { FC, PropsWithChildren, useMemo } from 'react';
import classNames from 'classnames';

import { ButtonGroupContext } from './button-group-context';
import {
  ButtonGroupProps,
  ButtonGroupConfig,
  ButtonGroupVariant,
} from './ButtonGroup.types';
import { useButtonGroupCSS } from './ButtonGroup.styles';

const buttonGroupVariants: ButtonGroupVariant[] = [
  'default',
  'outline',
  'ghost',
];

const ButtonGroup: FC<PropsWithChildren<ButtonGroupProps>> = ({
  size = 'md',
  type = 'default',
  variant = 'default',
  disabled = false,
  children,
  className = '',
  vertical = false,
  ...resetProps
}) => {
  const buttonGroupVariant: ButtonGroupVariant = buttonGroupVariants.includes(
    variant
  )
    ? variant
    : 'default';
  const initialConfig = useMemo<ButtonGroupConfig>(
    () => ({
      size,
      type,
      variant: buttonGroupVariant,
      disabled,
      isButtonGroup: true,
    }),
    [disabled, size, type, buttonGroupVariant]
  );

  const { className: resolveClassName, styles } = useButtonGroupCSS({
    type,
    size,
    variant: buttonGroupVariant,
    disabled,
  });

  const classes = classNames(
    'raw-button-group',
    vertical && 'raw-button-group-vertical',
    !vertical && 'raw-button-group-horizontal',
    className,
    resolveClassName
  );

  return (
    <ButtonGroupContext.Provider value={initialConfig}>
      <div role="group" className={classes} {...resetProps}>
        {children}
        {styles}
      </div>
    </ButtonGroupContext.Provider>
  );
};

export default ButtonGroup;
