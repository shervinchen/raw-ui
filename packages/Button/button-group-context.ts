import { createContext, useContext } from 'react';
import { ButtonGroupConfig } from './ButtonGroup.types';

const defaultContext: ButtonGroupConfig = {
  size: 'md',
  type: 'default',
  variant: 'default',
  disabled: false,
  isButtonGroup: false,
};

export const ButtonGroupContext =
  createContext<ButtonGroupConfig>(defaultContext);

export const useButtonGroupContext = (): ButtonGroupConfig =>
  useContext<ButtonGroupConfig>(ButtonGroupContext);
