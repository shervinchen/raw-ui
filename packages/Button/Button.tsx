import React, {
  forwardRef,
  Children,
  MouseEvent,
  useRef,
  useImperativeHandle,
  ComponentPropsWithRef,
} from 'react';
import classNames from 'classnames';

import { useButtonStyles, useButtonCSS } from './Button.styles';
import { ButtonProps } from './Button.types';
import ButtonLoading from './ButtonLoading';
import ButtonIcon from './ButtonIcon';
import { useButtonGroupContext } from './button-group-context';
import { ButtonGroupConfig } from './ButtonGroup.types';

const mergeButtonGroupProps = (
  buttonProps: ButtonProps,
  config: ButtonGroupConfig
): ButtonProps => {
  if (!config.isButtonGroup) return buttonProps;
  return {
    ...buttonProps,
    size: config.size,
    type: config.type,
    variant: config.variant,
    disabled: config.disabled,
  };
};

const Button = forwardRef(
  (buttonProps: ButtonProps, ref: ComponentPropsWithRef<'button'>['ref']) => {
    const buttonGroupConfig = useButtonGroupContext();
    const {
      className = '',
      size = 'md',
      type = 'default',
      variant = 'default',
      htmlType = 'button',
      loading = false,
      disabled = false,
      icon,
      iconRight,
      onClick,
      children,
      ...restProps
    } = mergeButtonGroupProps(buttonProps, buttonGroupConfig);
    const buttonRef = useRef<HTMLButtonElement>(null);

    const { horizontalPadding, height, color, backgroundColor } =
      useButtonStyles({
        type,
        size,
        variant,
        loading,
        disabled,
      });

    const { className: resolveClassName, styles } = useButtonCSS({
      type,
      size,
      variant,
      loading,
      disabled,
    });

    const isChildLess = Children.count(children) === 0;
    const isRight = Boolean(iconRight);

    const classes = classNames(
      'raw-button',
      loading && 'raw-loading-button',
      disabled && 'raw-disabled-button',
      isChildLess && 'raw-childless-button',
      className,
      resolveClassName
    );

    const handleClick = (event: MouseEvent<HTMLButtonElement>) => {
      if (loading || disabled) return;
      onClick?.(event);
    };

    useImperativeHandle(ref, () => buttonRef.current);

    return (
      <button
        ref={buttonRef}
        type={htmlType}
        className={classes}
        onClick={handleClick}
        disabled={disabled}
        aria-disabled={disabled}
        {...restProps}
      >
        {loading && (
          <ButtonLoading color={color} backgroundColor={backgroundColor} />
        )}
        {icon && (
          <ButtonIcon
            isSingle={isChildLess}
            height={height}
            horizontalPadding={horizontalPadding}
          >
            {icon}
          </ButtonIcon>
        )}
        {!isChildLess && (
          <span className="raw-button-content" data-testid="buttonContent">
            {children}
          </span>
        )}
        {iconRight && (
          <ButtonIcon
            isSingle={isChildLess}
            isRight={isRight && !isChildLess}
            height={height}
            horizontalPadding={horizontalPadding}
          >
            {iconRight}
          </ButtonIcon>
        )}
        {styles}
      </button>
    );
  }
);

Button.displayName = 'RawButton';

export default Button;
