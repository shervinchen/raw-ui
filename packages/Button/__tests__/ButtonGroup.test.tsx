import React from 'react';
import { render, screen } from '@testing-library/react';
import Button from '..';
import { ButtonSizes, ButtonTypes } from '../Button.types';
import { ButtonGroupVariant } from '../ButtonGroup.types';

const typeColorMap = {
  primary: 'rgb(0, 0, 0)',
  success: 'rgb(0, 112, 243)',
  warning: 'rgb(245, 166, 35)',
  error: 'rgb(238, 0, 0)',
};

const sizeHeightMap = {
  sm: '34px',
  md: '40px',
  lg: '46px',
};

describe('ButtonGroup', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(
      <Button.Group>
        <Button>One</Button>
        <Button>Two</Button>
        <Button>Three</Button>
      </Button.Group>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should support custom class name', () => {
    render(
      <Button.Group className="custom-button-group">
        <Button>One</Button>
        <Button>Two</Button>
        <Button>Three</Button>
      </Button.Group>
    );
    expect(screen.getByRole('group')).toHaveClass('custom-button-group');
  });

  test('should render the correct number of child', () => {
    render(
      <Button.Group>
        <Button>One</Button>
        <Button>Two</Button>
        <Button>Three</Button>
      </Button.Group>
    );
    expect(screen.getAllByRole('button').length).toBe(3);
  });

  ['primary', 'success', 'warning', 'error'].forEach(
    (item: Exclude<ButtonTypes, 'default'>) => {
      test(`should render ${item} type`, () => {
        render(
          <Button.Group type={item}>
            <Button>One</Button>
            <Button>Two</Button>
            <Button>Three</Button>
          </Button.Group>
        );
        const buttons = screen.getAllByRole('button');
        buttons.forEach((button) => {
          expect(button).toHaveStyle(`background-color: ${typeColorMap[item]}`);
        });
      });
    }
  );

  ['sm', 'md', 'lg'].forEach((item: ButtonSizes) => {
    test(`should render ${item} size`, () => {
      render(
        <Button.Group size={item}>
          <Button>One</Button>
          <Button>Two</Button>
          <Button>Three</Button>
        </Button.Group>
      );
      const buttons = screen.getAllByRole('button');
      buttons.forEach((button) => {
        expect(button).toHaveStyle(`height: ${sizeHeightMap[item]}`);
      });
    });
  });

  test('should render outline variant', () => {
    render(
      <Button.Group type="primary" variant="outline">
        <Button>One</Button>
        <Button>Two</Button>
        <Button>Three</Button>
      </Button.Group>
    );
    const buttons = screen.getAllByRole('button');
    buttons.forEach((button) => {
      expect(button).toHaveStyle(`color: rgb(0, 0, 0)`);
    });
  });

  test('should render ghost variant', () => {
    render(
      <Button.Group type="primary" variant="ghost">
        <Button>One</Button>
        <Button>Two</Button>
        <Button>Three</Button>
      </Button.Group>
    );
    const buttons = screen.getAllByRole('button');
    buttons.forEach((button) => {
      expect(button).toHaveStyle(`background-color: transparent`);
    });
  });

  test('should get default style when variant is unknown', () => {
    render(
      <Button.Group variant={'unknown' as ButtonGroupVariant}>
        <Button>One</Button>
        <Button>Two</Button>
        <Button>Three</Button>
      </Button.Group>
    );
    const buttons = screen.getAllByRole('button');
    buttons.forEach((button) => {
      expect(button).toHaveStyle(`background-color: rgb(255, 255, 255)`);
    });
  });

  test('should support vertical', () => {
    render(
      <Button.Group vertical>
        <Button>One</Button>
        <Button>Two</Button>
        <Button>Three</Button>
      </Button.Group>
    );
    expect(screen.getByRole('group')).toHaveClass('raw-button-group-vertical');
  });

  test('should support disabled', () => {
    render(
      <Button.Group disabled>
        <Button>One</Button>
        <Button>Two</Button>
        <Button>Three</Button>
      </Button.Group>
    );
    const buttons = screen.getAllByRole('button');
    buttons.forEach((button) => {
      expect(button).toBeDisabled();
    });
  });
});
