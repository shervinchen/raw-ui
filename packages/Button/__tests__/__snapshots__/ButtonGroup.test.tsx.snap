// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="raw-button-group raw-button-group-horizontal jsx-2328814491"
    role="group"
  >
    <button
      aria-disabled="false"
      class="raw-button jsx-2168746361"
      type="button"
    >
      <span
        class="raw-button-content"
        data-testid="buttonContent"
      >
        One
      </span>
    </button>
    <button
      aria-disabled="false"
      class="raw-button jsx-2168746361"
      type="button"
    >
      <span
        class="raw-button-content"
        data-testid="buttonContent"
      >
        Two
      </span>
    </button>
    <button
      aria-disabled="false"
      class="raw-button jsx-2168746361"
      type="button"
    >
      <span
        class="raw-button-content"
        data-testid="buttonContent"
      >
        Three
      </span>
    </button>
  </div>
</DocumentFragment>
`;
